File: app/src/main/java/com/chess/cryptobot/model/response/livecoin/LivecoinOrdersResponse.java
package com.chess.cryptobot.model.response.livecoin;

import com.chess.cryptobot.model.response.HistoryResponse;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.List;

public class LivecoinOrdersResponse extends LivecoinResponse {

    @SerializedName("data")
    @Expose
    private List<Datum> data = null;

    public List<Datum> getData() {
        return data;
    }

    public class Datum implements HistoryResponse {

        @SerializedName("currencyPair")
        @Expose
        public String currencyPair;
        @SerializedName("type")
        @Expose
        public String type;
        @SerializedName("issueTime")
        @Expose
        public Long issueTime;
        @SerializedName("price")
        @Expose
        public Double price;
        @SerializedName("quantity")
        @Expose
        public Double quantity;
        @SerializedName("remainingQuantity")
        @Expose
        public Double remainingQuantity;

        @Override
        public LocalDateTime getHistoryTime() {
            return LocalDateTime.ofEpochSecond(Math.round(issueTime / 1000), 0,
                    ZoneOffset.systemDefault().getRules().getOffset(Instant.now()));
        }

        @Override
        public String getHistoryName() {
            String[] currencies = currencyPair.split("/");
            return currencies[1].concat("/").concat(currencies[0]);
        }

        @Override
        public String getHistoryMarket() {
            return "livecoin";
        }

        @Override
        public Double getHistoryAmount() {
            return quantity;
        }

        @Override
        public Double getHistoryPrice() {
            return price;
        }

        @Override
        public String getHistoryAction() {
            return type;
        }

        @Override
        public Integer getProgress() {
            if (quantity==0d) return 0;
            return Double.valueOf(((quantity - remainingQuantity) / quantity) * 100).intValue();
        }
    }
}

File: app/src/main/java/com/chess/cryptobot/model/response/livecoin/LivecoinOrdersResponse.kt
package com.chess.cryptobot.model.response.livecoin

import com.chess.cryptobot.model.response.HistoryResponse
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import java.time.*
import kotlin.math.roundToLong

class LivecoinOrdersResponse : LivecoinResponse() {
    @SerializedName("data")
    @Expose
    val data: List<Datum>? = null

    inner class Datum : HistoryResponse {
        @SerializedName("currencyPair")
        @Expose
        var currencyPair: String? = null

        @SerializedName("type")
        @Expose
        override var historyAction: String? = null

        @SerializedName("issueTime")
        @Expose
        var issueTime: Long? = null

        @SerializedName("price")
        @Expose
        override var historyPrice: Double? = null

        @SerializedName("quantity")
        @Expose
        override var historyAmount: Double? = null

        @SerializedName("remainingQuantity")
        @Expose
        var remainingQuantity: Double? = null
        override val historyTime: ZonedDateTime?
            get() = ZonedDateTime.ofLocal(LocalDateTime.ofEpochSecond((issueTime!! / 1000.toFloat()).roundToLong(),
                    0,
                    ZoneOffset.UTC),
                    ZoneId.of("Z"), ZoneOffset.UTC)

        override val historyName: String
            get() {
                val currencies: Array<String> = currencyPair!!.split("/".toRegex()).toTypedArray()
                return currencies[1] + "/" + currencies[0]
            }

        override val historyMarket: String
            get() {
                return "livecoin"
            }

        override val progress: Int
            get() {
                if (historyAmount == 0.0) return 0
                return (((historyAmount!! - (remainingQuantity)!!) / (historyAmount)!!) * 100).toInt()
            }
    }
}